src/simple_key_provider.rs

```diff
@@ -11,6 +11,69 @@ use std::cell::RefCell;
 use crate::errors::{KeyDecryptionError, KeyGenerationError};
 use crate::key_provider::{DataKey, KeyProvider};
 
+// EncryptedSimpleKey relies on this size being constant. If this ever needs to be changed a new
+// version of EncryptedSimpleKey needs to be created.
+type Nonce = aes_gcm::Nonce<U12>;
+const NONCE_SIZE: usize = 12;
+
+/// A decoded intermediate representation of an encrypted simple key
+///
+/// The encoded version of the encrypted simple key looks like so:
+///
+/// | Pos  | Data                   |
+/// | -----|------------------------|
+/// | 0    | Version tag (1 byte)   |
+/// | 1-13 | Nonce       (12 bytes) |
```

Comment:
A 12-byte nonce is the bare minimum. It would be preferred to use a 16-byte Nonce but not sure if there is a `Nonce` type to support this. It might have to be `Nonce16 = GenericArray<u8, U16>` or something.

Reply:
We're using `Aes128Gcm` at the moment which is a type alias for `AesGcm<Aes128, U12>` but I can change it to use `AesGcm<Aes128, U16>`. I have no clue what the implications are either way so I'll update it to use `U16`. 

Not sure if it's relevant but this is a comment on the `AesGcm` struct:

![image](https://user-images.githubusercontent.com/17743045/170165624-532f2e36-c49e-43c5-97ee-9bc4585136db.png)

Reply:
The 96-bit nonce is based on the NIST guidelines which is sufficient for most applications. However the nonce size dictates how often you should rotate keys. Given that we a *lot* of data could be encrypted using envelopers a 128-bit nonce means keys would need to be rotated much less frequently. Its not an issue with KMS because KMS autorotates keys.

For reference, a 96-bit nonce means keys should be rotated after ~2^28 encryptions (2 GB of data).
A 128-bit nonce should be rotated after ~2^40 encryptions (16TB of data).